flake.nix

```diff
@@ -11,7 +11,7 @@
   # Once this is updated to something containing
   # https://github.com/NixOS/nixpkgs/pull/271423, don't forget
   # to remove the `nix.checkAllErrors = false;` line in the tests.
-  inputs.nixpkgs.url = "github:NixOS/nixpkgs/staging-23.05";
+  inputs.nixpkgs.url = "github:9999years/nixpkgs/update-tests-for-nix-9606-staging-23.05";
```

Comment:
This can be removed when https://github.com/NixOS/nixpkgs/pull/274077 is merged, just a hack to get tests passing.

The `lib` test suite has a few test cases that check the output of `nix-instantiate --eval-only`. I don't think that output is guaranteed by any compatibility policy, so I think we're good to change it, but we do need to patch the tests. Once we have the output finalized here, we can merge this PR in nixpkgs and then update the Flake input reference and remove this patch.

Reply:
For some reason updating the `nixpkgs` input makes `nix flake check` want to rebuild LLVM and GHC? So that's hitting the 1hr CI timeout.

Reply:
> I don't think that output is guaranteed by any compatibility policy

I would *feel* the same, but in Nix that thinking tends to lead to problems for users.

We're currently already interacting with Nixpkgs a bit, so that makes merging this line a bit hard.
To get this merged, I'd prefer to keep the "ascii" syntax around, perhaps through a bool flag?

Reply:
> rebuild LLVM and GHC

I would expect this to be resolved now, but I'm surprised that this happened. We do care about avoiding this.

Reply:
I'm beginning to think the test suite should just use the JSON output instead. Also an easy change to make: https://github.com/NixOS/nixpkgs/pull/275264

Reply:
Any machine parsing like the test suite should *definitely* use the JSON output. If the output of human-interface CLI tools shouldn't be changed to improve readability then that defeats the purpose of having CLI tools

Reply:
And more broadly, if we want to be able to actually improve the experience of using Nix, we have to be comfortable saying "no, that's not part of the public interface, you'll need to change your code because we haven't made a guarantee to keep this stable and writing a shell script that depends on it doesn't change that".

It sounds like there _is_ need for a stable output layer that can encode (e.g.) thunks and functions, so maybe we should add a second JSON output layer that has an actual public schema that we can stick to.

We'd just add another layer to disambiguate types, so we'd encode a thunk as `{"type": "thunk"}` (possibly with auxiliary information like where it comes from, if it's a cycle, etc.), and we could encode an attrset as `{"type": "attrset", "values": {...}}`. This would be machine-readable (without writing a Nix parser!) and easy to keep stable.

But in the meantime there don't exist any use-cases that can't be easily patched by using the JSON output and (for checking laziness) checking the error message for "cannot convert thunk to JSON".

Reply:
Marking this resolved as we got https://github.com/NixOS/nixpkgs/pull/275264 merged and I updated the Nixpkgs input, so we're all good to release now.

Reply:
If I remember correctly: This is annoying for the Tvix people (cc @sternenseemann) because they reuse the Nix language test suite as a compatibility check, and that test suite uses the result of `nix-instantiate --eval`.

I don't think that's really _right_ (as you said, it's not great to rely on that output), but since it is the _de facto_ canonical language test, there's not a lot of leeway.

All that to say, changing the pretty printer would probably require changing the test suite first to not rely on the pretty printing (but use the XML or JSON format) first.